File: questModule/src/main/java/ru/nekit/android/qls/quest/types/PerimeterQuest.java
package ru.nekit.android.qls.quest.types;

import android.content.Context;
import android.support.annotation.NonNull;

import ru.nekit.android.qls.R;
import ru.nekit.android.qls.quest.QuestType;
import ru.nekit.android.qls.quest.common.Quest;

public class PerimeterQuest extends NumberSummandQuest {

    public PerimeterQuest(@NonNull Quest quest) {
        setQuestType(QuestType.PERIMETER);
        setQuestionType(quest.getQuestionType());
        NumberSummandQuest inQuest = (NumberSummandQuest) quest;
        int[] values = new int[2];
        values[0] = inQuest.leftNode[0];
        values[1] = inQuest.leftNode[1];
        if (values[1] == 0) {
            values[1] = values[0];
        }
        leftNode = values;
        unknownMemberIndex = inQuest.unknownMemberIndex;
        unknownOperatorIndex = inQuest.unknownOperatorIndex;
        if (quest.getAvailableAnswerVariants() != null) {
            setAvailableAnswerVariants(quest.getAvailableAnswerVariants());
        }
    }

    public String getFigureName(@NonNull Context context) {
        return context.getString(isSquare() ? R.string.square_figure_name : R.string.rectangle_figure_name);
    }

    public boolean isSquare() {
        return getASideSize() == getBSideSize();
    }

    public int getASideSize() {
        return leftNode[0];
    }

    public int getBSideSize() {
        return leftNode[1];
    }

    public int getPerimeter() {
        return (getASideSize() + getBSideSize()) * 2;
    }

    @Override
    public Integer getAnswer() {
        switch (getQuestionType()) {

            case SOLUTION:

                return getPerimeter();

        }
        return super.getAnswer();
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/quest/types/PerimeterQuest.kt
package ru.nekit.android.qls.quest.types

import android.content.Context

import ru.nekit.android.qls.R
import ru.nekit.android.qls.domain.model.quest.NumberSummandQuest
import ru.nekit.android.qls.domain.model.quest.Quest
import ru.nekit.android.qls.shared.model.QuestType
import ru.nekit.android.qls.shared.model.QuestionType

//ver 1.0
class PerimeterQuest(quest: Quest) : NumberSummandQuest() {

    val isSquare: Boolean
        get() = aSideSize == bSideSize

    val aSideSize: Int
        get() = leftNode[0]

    val bSideSize: Int
        get() = leftNode[1]

    val perimeter: Int
        get() = (aSideSize + bSideSize) * 2

    override val answer: Int
        get() {
            when (questionType) {

                QuestionType.SOLUTION ->

                    return perimeter
            }
            return super.answer
        }

    init {
        questType = QuestType.PERIMETER
        questionType = quest.questionType
        val inQuest = quest as NumberSummandQuest
        val values = IntArray(2)
        values[0] = inQuest.leftNode[0]
        values[1] = inQuest.leftNode[1]
        if (values[1] == 0) {
            values[1] = values[0]
        }
        leftNode = values
        unknownMemberIndex = inQuest.unknownMemberIndex
        unknownOperatorIndex = inQuest.unknownOperatorIndex
        if (quest.availableAnswerVariants != null) {
            availableAnswerVariants = quest.availableAnswerVariants
        }
    }

    fun getFigureName(context: Context): String {
        return context.getString(if (isSquare) R.string.square_figure_name else R.string.rectangle_figure_name)
    }
}